### DISCUSSION TITLE
How to replicate the standard textarea behaviour for newlines?

### DISCUSSION THREAD

**Discussion by emcarson** [livewire/livewire#2483] [1. Help] [answered]
Hello all,

I've recently converted my standard HTML forms to Livewire v1.

With my standard forms, all newlines were converted to "CRLF" ("\r\n") automatically (according to HTML standards), no matter the user's OS, browser, etc.

I noticed that in my case (Linux), Livewire is sending the newlines just as "\n" (which is a problem, since this content is later uploaded to an API that expects "\r\n"). 

Is this behaviour expected?

I could replace all "\n" to "\r\n" manually, but isn't there a chance that I'll receive "\r\n" from a Windows user and "\n" from a Linux user? How could I handle this?

Thank you for your advice.

**Comment by joshhanley** [comment, marked as answer]
@emcarson you could probably use an `updated` hook method to review all data and do a check to ensure it's the correct format. Or you could just process the data on save, so it's ready for upload to the API.

You would probably need to use string helpers or regex to check if it ends in `\r\n` or just `\n` and fix it if required.

Have a look here for `updated` hook https://laravel-livewire.com/docs/2.x/lifecycle-hooks

Hope this helps!

**Reply by emcarson** [reply]
Thank you very much @joshhanley, it did help!

I ended up creating a Trait to replicate Laravel's behaviour as much as possible (trim strings automatically, convert empty strings to null, fix newlines, etc).

I'll just throw it in here in case it's useful to someone:

```php
<?php

namespace App\Traits\Components;

use Exception;

/**
 * Trait TrimsAndNullifies
 * @property array $except
 */
trait TrimsAndNullifies
{
    /**
     * Get the cleaning exceptions.
     *
     * @return array
     * @throws \Exception
     */
    protected function getExcept()
    {
        if (! property_exists($this, 'except') || ! is_array($this->except)) {
            throw new Exception('An [except] property must be defined on [' . get_class($this) . ']');
        }

        return $this->except;
    }

    /**
     * Cleans an input array.
     *
     * @param  array  $input
     * @return array
     */
    protected function cleanInputArray($input)
    {
        array_walk($input, function(&$value, $key) {
            $value = $this->cleanInput($key, $value);
        });

        return $input;
    }

    /**
     * Cleans a specific input.
     *
     * @param  string  $field
     * @param  mixed  $value
     * @return mixed
     */
    protected function cleanInput($field, $value)
    {
        if (in_array($field, $this->getExcept(), true)) {
            return $value;
        }

        if (is_string($value)) {

            $regex = '~(?<!\\r)\\n~';
            $value = preg_replace($regex, "\r\n", $value);

            $value = trim($value);

            return ($value === '') ? null : $value;
        }

        return $value;
    }
}
````
